File: leetcode/0337robbers3/my_answer.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def rob(self, root: TreeNode) -> int:
        from collections import defaultdict
        seen = defaultdict(int)
        self.robHelp(root, seen)
        return seen[root]
        
    def robHelp(self, root: TreeNode, seen) -> int:
        if not root:
            return 0
        if seen[root] != 0:
            return seen[root]
        
        r = root.val
        if root.right:
            r += self.robHelp(root.right.left, seen) + self.robHelp(root.right.right, seen)
        if root.left:
            r += self.robHelp(root.left.left, seen) + self.robHelp(root.left.right, seen)
        
        seen[root] = max(r, self.robHelp(root.left, seen) + self.robHelp(root.right, seen))
        return seen[root]

File: leetcode/0236/my_answer.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* r;
    
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        this->r = NULL;
        this->helper(root, p, q);
        return this->r;
    }
    
    vector<bool> helper(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (root == NULL) return {false, false};
        vector<bool> l = helper(root->left, p, q);
        vector<bool> r = helper(root->right, p, q);
        vector<bool> a = {l[0] || r[0] || p == root, l[1] || r[1] || q == root};
        if (a[0] && a[1] && !this->r) this->r = root;
        return a;
    }
};
